lib/Sema/CSFix.cpp

```diff
@@ -759,6 +759,44 @@ IgnoreContextualType *IgnoreContextualType::create(ConstraintSystem &cs,
       IgnoreContextualType(cs, resultTy, specifiedTy, locator);
 }
 
+bool RemoveUnnecessaryCoercion::diagnose(Expr *root, bool asNote) const {
+  auto &cs = getConstraintSystem();
+  UnnecessaryCoercionFailure failure(root, cs, getFromType(), getToType(),
+                                     getLocator());
+  return failure.diagnose(asNote);
+}
+
+bool
+RemoveUnnecessaryCoercion::attempt(ConstraintSystem &cs,
+                                   Type fromType,
+                                   Type toType,
+                                   ConstraintLocatorBuilder locator) {
+  if (auto last = locator.last()) {
+    if (last->is<LocatorPathElt::ExplicitTypeCoercion>()) {
+      
+      auto expr = cast<CoerceExpr>(locator.getAnchor());
+      auto toTypeRepr = expr->getCastTypeLoc().getTypeRepr();
+      
+      // only diagnosing for coercion where the left-side is a DeclRefExpr
+      // or a explicit/implicit coercion e.g. Double(1) as Double
+      if (!isa<ImplicitlyUnwrappedOptionalTypeRepr>(toTypeRepr) &&
+          (isa<DeclRefExpr>(expr->getSubExpr()) || isa<CoerceExpr>(expr->getSubExpr()))) {
+        
+        auto *locatorPtr = cs.getConstraintLocator(locator);
+        // If the constraint system has a baseCS here, it's probably coming from
+        // a re-type check and this fix was already emitted, so we skip it here.
+        if (!cs.hasFixFor(locatorPtr) && !cs.baseCS) {
```

Comment:
I'd say instead of assuming that there are duplicate fixes like that here, it would be better check that while emitting a diagnostic, at that point constraint system would have all of the fixes which were necessary for given solution.

Reply:
Hey @xedin :)
One question about that, is at this point it will have all the fixes for an instance of a constraint system, right? So we would check for duplicated fixes on this constraint system? But that would not solve the issue here because there's one fix of kind RemoveUnnecessaryCoercion on cs.fixes when we are emitting the diagnostics on `UnnecessaryCoercionFailure::diagnoseAsError`. The duplication is because when a nested coercion fails and when trying to diagnose the failure it ends up re-type checking, which at some point in this process creates a new constraint system that ends up also having a fixit too. So the dupe fixes are in two different systems as I mentioned in the comment before, so I'm not sure how I could implement this check there. The only thing I could think is to check and skip emitting the diag if the cs has a baseCS && this baseCS also has a fix of kind RemoveUnnecessaryCoercion because this second cs has the first as the base, would that be correct?

Reply:
Also, @xedin if we check this there on `UnnecessaryCoercionFailure::diagnoseAsError` and return false if it's a duplication, we will end up hitting an assertion of all warnings should be diagnosed. So if we move the check there should we always return true?

Reply:
Sorry I don't quite understand what you are trying to say, do you have a concrete example to demonstrate how this would happen? I'll try to pull your changes in and run under debugger to understand it better. I thought that it's something like `1 as Int as Int` or something but I don't understand how we end up in CSDiag. What throws me off is mention that "this fix was already emitted", I'm not sure how can we know that is true.

Reply:
Oh sorry @xedin,
There are few cases, but a simple one would be,
```swift
Double(1) as Double as String

```
Note that there is a failure on coercion (double as string), the system fails, tries to salvage and it ends up [here](https://github.com/apple/swift/blob/4ade7ae11f628b5f33d51de8e90489aa45039a2b/lib/Sema/CSDiag.cpp#L3699), so in this re-type checking the child sub (`Double(1) as Double`) it creates a new system that records the fix it again too. As far as I could understand this re-type checking is done to attempt to provide better diagnostics.

Reply:
Also, I'm not sure if this is the correct check, but the intention is to skip record this if it was already typechecked, meaning it probably already emitted this. It's not a commom case, I think in the whole test suit it triggered duplicated in two places.

Reply:
It kind of sounds like tweak to existing `as to as!` fix to recognize/diagnose invalid explicit coercions is going to resolve that problem because we are no longer going to fall-through into CSDiag...

Reply:
> It kind of sounds like tweak to existing `as to as!` fix to recognize/diagnose invalid explicit coercions is going to resolve that problem because we are no longer going to fall-through into CSDiag...

@xedin, I've implemented on last commit so it solved the duplication for the case above. There is only one edge case left to handle where there is a duplication, that is 
```swift
func f(_ ns: NSString?) {
  var s2 = ns ?? "str" as String as String // expected-error {{cannot convert value of type 'String' to expected argument type 'NSString'}}{{43-43= as NSString}} 
}
```
Although this argument conversion failure is diagnosed via a fix already, for some reason is still going down to CSDiag and only emitting the fix when re-typechecking, still not sure why it happens. 
But besides that, do you think there is anything else to do in this PR overall or this is almost good to go?

Reply:
I think it's okay to leave that to report multiple warnings for now, but please add a `TODO(diagnostic)` to the suite to identify that's a known issue. The reason why it happens we fail to fix the original constraint system but CSDiag then manages to find a fix for one of the sub-expressions. Looks like we just need to add force downcast fix for locators like `[Binary -> apply argument -> comparing call argument #1 to parameter #1 -> @autoclosure result]]]`

Reply:
This is simple to solve? maybe now that I'm going to open a new PR for the ForceDownCast, I can handle this one there too? I was trying to emit a fix for that on `matchTypes` when the constraint kind is `ConstraintKind::OperatorArgumentConversion:` but that was not fixing the issue, didn't spend much time there.

Reply:
If it's not so simple, I can just add the `TODO(diagnostic)` as you mention and open a third PR after this land, what you think?

Reply:
I think adding a `TODO(diagnostics)` for it and addressing it separately is okay.